fish-rust/src/builtins/return_exit.rs

```diff
@@ -0,0 +1,148 @@
+use crate::wutil::fish_wcstoi;
+use crate::wutil::wgettext_fmt;
+
+#[derive(Debug, Clone, Copy)]
+struct Options {
+    print_help: bool,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self { print_help: false }
+    }
+}
+
+fn parse_options(
+    args: &mut [&wstr],
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+) -> Result<(Options, usize), Option<c_int>> {
+    let cmd = args[0];
+
+    const SHORT_OPTS: &wstr = L!(":h");
+    const LONG_OPTS: &[woption] = &[wopt(L!("help"), woption_argument_t::no_argument, 'h')];
+
+    let mut opts = Options::default();
+
+    let mut w = wgetopter_t::new(SHORT_OPTS, LONG_OPTS, args);
+
+    while let Some(c) = w.wgetopt_long() {
+        match c {
+            'h' => opts.print_help = true,
+            ':' => {
+                builtin_missing_argument(parser, streams, cmd, args[w.woptind - 1], true);
+                return Err(STATUS_INVALID_ARGS);
+            }
+            '?' => {
+                // We would normally invoke builtin_unknown_option() and return an error.
+                // But for this command we want to let it try and parse the value as a negative
+                // return value.
+                return Ok((opts, w.woptind - 1));
+            }
+            _ => {
+                panic!("unexpected retval from wgetopt_long");
+            }
+        }
+    }
+
+    Ok((opts, w.woptind))
+}
+
+/// Function for handling the return builtin.
+pub fn r#return(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    args: &mut [&wstr],
+) -> Option<c_int> {
+    let mut retval = match parse_return_value(args, parser, streams) {
+        Ok(v) => v,
+        Err(e) => return e,
+    };
+
+    let has_function_block = parser.ffi_has_funtion_block();
+
+    // *nix does not support negative return values, but our `return` builtin happily accepts being
+    // called with negative literals (e.g. `return -1`).
+    // Map negative values to (256 - their absolute value). This prevents `return -1` from
+    // evaluating to a `$status` of 0 and keeps us from running into undefined behavior by trying to
+    // left shift a negative value in W_EXITCODE().
+    if retval < 0 {
+        retval = 256 - (abs(retval) % 256);
+    }
+
+    // If we're not in a function, exit the current script (but not an interactive shell).
+    if !has_function_block {
+        if !parser.is_interactive() {
+            parser.pin().libdata().set_exit_current_script(true);
+        }
+        return Some(retval);
+    }
+
+    // Mark a return in the libdata.
+    parser.pin().libdata().set_returning(true);
+
+    return Some(retval);
+}
+
+/// Function for handling the exit builtin.
+pub fn exit(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    args: &mut [&wstr],
+) -> Option<c_int> {
+    let retval =  match parse_return_value(args, parser, streams) {
+        Ok(v) => v,
+        Err(e) => return e,
+    };
+
```

Comment:
Missing comment from original code:

```suggestion

    // Mark that we are exiting in the parser.
    // TODO: in concurrent mode this won't successfully exit a pipeline, as there are other parsers
    // involved. That is, `exit | sleep 1000` may not exit as hoped. Need to rationalize what
    // behavior we want here.
```